File: backend/trilateration.py

import numpy as np
from scipy.optimize import least_squares

class TrilaterationCalculator:
    def __init__(self):
        # Known positions of landmarks (blocks) in meters
        # Using Block A as reference point (0,0)
        self.landmark_positions = {
            "Block A": np.array([0, 0]),      # Reference point
            "Block B": np.array([-75.32, -54.60]),   # Calculated from your GeoJSON
            "Block C": np.array([-96.63, -90.62]),
            "Block D": np.array([-3.91, -103.76]),
            "Block E": np.array([14.81, -51.15]),
            "Block F": np.array([52.91, -159.41])
        }

    def calculate_position_with_confidence(self, measurements):
        """
        Calculate position with confidence estimates
        
        Args:
            measurements: List of dicts with format:
                [{
                    'block': 'Block A',
                    'distance': 10.5,  # in meters
                    'confidence': 0.95  # optional confidence score
                }, ...]
        
        Returns:
            {
                'position': [x, y],
                'error_estimate': float,
                'used_landmarks': int
            }
        """
        try:
            # Filter and weight measurements
            valid_measurements = []
            weights = []
            
            for m in measurements:
                if m['block'] in self.landmark_positions:
                    valid_measurements.append(m)
                    # Use confidence if available, otherwise default to 1.0
                    weights.append(m.get('confidence', 1.0))
            
            if len(valid_measurements) < 3:
                raise ValueError("Not enough valid measurements (need at least 3)")
            
            positions = [self.landmark_positions[m['block']] for m in valid_measurements]
            distances = [m['distance'] for m in valid_measurements]
            
            # Weighted initial guess (centroid of landmarks)
            initial_guess = np.average(positions, axis=0, weights=weights)
            
            # Weighted error function
            def error_function(point):
                errors = [
                    (np.linalg.norm(point - positions[i]) - distances[i])
                    for i in range(len(positions))
                ]
                return np.array(errors) * np.array(weights)
            
            # Solve using nonlinear least squares
            result = least_squares(error_function, initial_guess)
            
            if not result.success:
                raise RuntimeError("Position calculation failed to converge")
            
            # Calculate error estimate (mean absolute residual)
            residuals = error_function(result.x)
            error_estimate = np.mean(np.abs(residuals))
            
            return {
                'position': result.x.tolist(),
                'error_estimate': error_estimate,
                'used_landmarks': len(valid_measurements)
            }
            
        except Exception as e:
            return {
                'error': str(e),
                'used_landmarks': len(valid_measurements) if 'valid_measurements' in locals() else 0
            }

    def latlng_to_meters(self, lat, lng, ref_lat=31.481991, ref_lng=74.3036737):
        """Convert latitude/longitude to meters relative to reference point"""
        # Approximate conversions for small areas
        lat_to_meters = 111320  # meters per degree latitude
        lng_to_meters = 111320 * np.cos(np.radians(ref_lat))  # meters per degree longitude
        
        x = (lng - ref_lng) * lng_to_meters
        y = (lat - ref_lat) * lat_to_meters
        return np.array([x, y])

File: src/screens/trilateration.js
// trilateration.js
import { Alert } from 'react-native';

// Convert latitude/longitude to meters (simplified for small campus area)
function toMeters(lat, lng, refLat = 31.481991, refLng = 74.3036737) {
  // Approximate conversions for small areas
  const latToMeters = 111320; // meters per degree latitude
  const lngToMeters = 111320 * Math.cos(refLat * Math.PI / 180); // meters per degree longitude
  
  return {
    x: (lng - refLng) * lngToMeters,
    y: (lat - refLat) * latToMeters
  };
}

// Block coordinates in meters (using your GeoJSON data)
const BLOCK_COORDS = {
  "Block A": toMeters(31.481991, 74.3036737),
  "Block B": toMeters(31.4815, 74.3030), // Approximate center of polygon
  "Block C": toMeters(31.4811762, 74.3028048),
  "Block D": toMeters(31.481064, 74.3033213),
  "Block E": toMeters(31.4815328, 74.3038079),
  "Block F": toMeters(31.4805619, 74.3041524)
};

export function calculatePosition(measurements) {
  // Need at least 3 measurements for trilateration
  if (measurements.length < 3) {
    throw new Error("Need at least 3 measurements for trilateration");
  }

  // Filter measurements for known blocks
  const validMeasurements = measurements.filter(m => BLOCK_COORDS[m.block]);
  
  if (validMeasurements.length < 3) {
    throw new Error("Need measurements to at least 3 known blocks");
  }

  // Prepare data for trilateration
  const points = validMeasurements.map(m => {
    const blockPos = BLOCK_COORDS[m.block];
    return {
      x: blockPos.x,
      y: blockPos.y,
      r: m.distance
    };
  });

  // Simple trilateration (least squares approach)
  let A = [], b = [];
  for (let i = 1; i < points.length; i++) {
    A.push([
      2 * (points[i].x - points[0].x),
      2 * (points[i].y - points[0].y)
    ]);
    b.push([
      points[i].x**2 + points[i].y**2 - points[i].r**2 - 
      (points[0].x**2 + points[0].y**2 - points[0].r**2)
    ]);
  }

  // Solve the system Ax = b
  const At = transpose(A);
  const AtA = multiply(At, A);
  const Atb = multiply(At, b);
  const solution = solve(AtA, Atb);

  if (!solution) {
    throw new Error("Could not solve trilateration equations");
  }

  return {
    x: solution[0][0],
    y: solution[1][0],
    error: calculateError(solution[0][0], solution[1][0], points)
  };
}

// Helper linear algebra functions
function transpose(matrix) {
  return matrix[0].map((_, i) => matrix.map(row => row[i]));
}

function multiply(a, b) {
  const result = [];
  for (let i = 0; i < a.length; i++) {
    result[i] = [];
    for (let j = 0; j < b[0].length; j++) {
      let sum = 0;
      for (let k = 0; k < a[0].length; k++) {
        sum += a[i][k] * b[k][j];
      }
      result[i][j] = sum;
    }
  }
  return result;
}

function solve(A, b) {
  // Simple Gaussian elimination for 2x2 system
  const a = A[0][0], b1 = A[0][1], c = A[1][0], d = A[1][1];
  const e = b[0][0], f = b[1][0];
  
  const det = a * d - b1 * c;
  if (Math.abs(det) < 1e-10) return null;
  
  return [
    [(d * e - b1 * f) / det],
    [(a * f - c * e) / det]
  ];
}

function calculateError(x, y, points) {
  // Calculate average error
  let totalError = 0;
  for (const p of points) {
    const dist = Math.sqrt((x - p.x)**2 + (y - p.y)**2);
    totalError += Math.abs(dist - p.r);
  }
  return totalError / points.length;
}